File: tools/sdks/swiftui/Sources/ThunderSwiftUI/components/presentation/auth/InviteUser.swift
import SwiftUI

/// Form to invite a user by email to the current organization (spec §8.4 Presentation).
public struct ThunderInviteUser: View {
    @EnvironmentObject private var i18n: ThunderI18n
    public let onSent: (() -> Void)?
    public let onError: ((String) -> Void)?

    public init(onSent: (() -> Void)? = nil, onError: ((String) -> Void)? = nil) {
        self.onSent = onSent
        self.onError = onError
    }

    public var body: some View {
        BaseThunderInviteUser(onSent: onSent, onError: onError) { email, isLoading, error, send in
            VStack(alignment: .leading, spacing: 12) {
                Text(i18n.resolve("inviteUser.title"))
                    .accessibilityAddTraits(.isHeader)
                if let error { Text(error).foregroundColor(.red) }
                TextField(i18n.resolve("inviteUser.email"), text: email)
                    .accessibilityLabel(i18n.resolve("inviteUser.email"))
                    .frame(minHeight: 44)
                Button(isLoading
                       ? i18n.resolve("inviteUser.loading")
                       : i18n.resolve("inviteUser.submit")
                ) { send() }
                    .disabled(isLoading)
                    .frame(minWidth: 44, minHeight: 44)
            }
            .padding()
        }
    }
}

/// Unstyled base variant (spec §8.3).
public struct BaseThunderInviteUser<Content: View>: View {
    @EnvironmentObject private var state: ThunderState
    public let onSent: (() -> Void)?
    public let onError: ((String) -> Void)?
    public let content: (Binding<String>, Bool, String?, () -> Void) -> Content

    @State private var email = ""
    @State private var isLoading = false
    @State private var error: String?

    public init(
        onSent: (() -> Void)? = nil,
        onError: ((String) -> Void)? = nil,
        @ViewBuilder content: @escaping (Binding<String>, Bool, String?, () -> Void) -> Content
    ) {
        self.onSent = onSent
        self.onError = onError
        self.content = content
    }

    public var body: some View {
        content($email, isLoading, error, send)
    }

    private func send() {
        guard !email.isEmpty else { return }
        isLoading = true
        error = nil
        Task {
            do {
                try await state.client.inviteUser(email: email)
                isLoading = false
                onSent?()
            } catch {
                self.error = error.localizedDescription
                isLoading = false
                onError?(error.localizedDescription)
            }
        }
    }
}

File: tools/sdks/compose/src/main/kotlin/io/thunder/compose/components/presentation/auth/InviteUser.kt
package io.thunder.compose.components.presentation.auth

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.BasicText
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import io.thunder.compose.LocalThunder
import io.thunder.compose.components.actions.BaseThunderSignUpButton
import kotlinx.coroutines.launch

/** Form to invite a user by email (spec §8.4 Presentation). */
@Composable
fun ThunderInviteUser(
    modifier: Modifier = Modifier,
    onSent: (() -> Unit)? = null,
    onError: ((String) -> Unit)? = null,
) {
    val thunderState = LocalThunder.current
    val i18n = thunderState.i18n
    BaseThunderInviteUser(modifier = modifier, onSent = onSent, onError = onError) { email, isLoading, error, send ->
        Column(modifier = Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
            BasicText(i18n.resolve("inviteUser.title"))
            error?.let { BasicText(it) }
            BasicTextField(
                value = email.value,
                onValueChange = { email.value = it },
                modifier = Modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp),
            )
            BaseThunderSignUpButton(
                label = if (isLoading) i18n.resolve("inviteUser.loading") else i18n.resolve("inviteUser.submit"),
                modifier = Modifier.fillMaxWidth(),
            ) { if (!isLoading) send() }
        }
    }
}

/** Unstyled base variant (spec §8.3). */
@Composable
fun BaseThunderInviteUser(
    modifier: Modifier = Modifier,
    onSent: (() -> Unit)? = null,
    onError: ((String) -> Unit)? = null,
    content: @Composable (email: MutableState<String>, isLoading: Boolean, error: String?, send: () -> Unit) -> Unit,
) {
    val thunderState = LocalThunder.current
    val scope = rememberCoroutineScope()
    val email = remember { mutableStateOf("") }
    var isLoading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    val send = {
        if (email.value.isNotBlank()) {
            scope.launch {
                isLoading = true; error = null
                try {
                    thunderState.client.inviteUser(email = email.value)
                    isLoading = false
                    onSent?.invoke()
                } catch (e: Exception) {
                    error = e.message
                    isLoading = false
                    onError?.invoke(e.message ?: "Failed to send invite")
                }
            }
        }
    }

    Box(modifier = modifier) { content(email, isLoading, error, send) }
}
